R-package/src/cmake/modules/FindLibR.cmake

```diff
@@ -0,0 +1,210 @@
+endif()
+
+# Creates R.lib and R.def in the build directory for linking with MSVC
+function(create_rlib_for_msvc)
+
+  message("Creating R.lib and R.def")
+
+  # various checks and warnings
+  if(NOT WIN32 OR NOT MSVC)
+    message(FATAL_ERROR "create_rlib_for_msvc() can only be used with MSVC")
+  endif()
+
+  if(NOT EXISTS "${LIBR_LIB_DIR}")
+    message(FATAL_ERROR "LIBR_LIB_DIR, '${LIBR_LIB_DIR}', not found")
+  endif()
+
+  find_program(GENDEF_EXE gendef)
+  find_program(DLLTOOL_EXE dlltool)
+
+  if(NOT GENDEF_EXE OR NOT DLLTOOL_EXE)
+    message(FATAL_ERROR "Either gendef.exe or dlltool.exe not found!\
+      \nDo you have Rtools installed with its MinGW's bin/ in PATH?")
+  endif()
+
+  # extract symbols from R.dll into R.def and R.lib import library
+  execute_process(COMMAND ${GENDEF_EXE}
+    "-" "${LIBR_LIB_DIR}/R.dll"
+    OUTPUT_FILE "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+  )
+  execute_process(COMMAND ${DLLTOOL_EXE}
+    "--input-def" "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+    "--output-lib" "${CMAKE_CURRENT_BINARY_DIR}/R.lib"
+  )
+endfunction(create_rlib_for_msvc)
+
+# R version information is used to search for R's libraries in
+# the registry on Windows. Since this code is orchestrated by
+# an R script (src/install.libs.R), that script uses R's built-ins to
+# find the version of R and pass it through as a CMake variable
+if(CMAKE_R_VERSION)
+  message("R version passed into FindLibR.cmake: ${CMAKE_R_VERSION}")
+elseif(WIN32)
+  message(FATAL_ERROR "Expected CMAKE_R_VERSION to be passed in on Windows but none was provided. Check src/install.libs.R")
+endif()
+
+
+if (NOT LIBR_EXECUTABLE)
+  find_program(
+    LIBR_EXECUTABLE
+    NAMES R R.exe
+  )
+
+  # CRAN may run RD CMD CHECK instead of R CMD CHECK,
+  # which can lead to this infamous error:
+  # 'R' should not be used without a path -- see par. 1.6 of the manual
+  if(LIBR_EXECUTABLE MATCHES ".*\\.Rcheck.*")
+    unset(LIBR_EXECUTABLE PARENT_SCOPE)
+    unset(LIBR_EXECUTABLE CACHE)
+  endif()
+
+  # ignore the R bundled with R.app on Mac, since that is GUI-only
+  if(LIBR_EXECUTABLE MATCHES ".+R\\.app.*")
+    unset(LIBR_EXECUTABLE PARENT_SCOPE)
+    unset(LIBR_EXECUTABLE CACHE)
+  endif()
+endif()
+
+# Find R executable unless it has been provided directly
+if (NOT LIBR_EXECUTABLE)
+  if(APPLE)
+
+    find_library(LIBR_LIBRARIES R)
+
+    if(LIBR_LIBRARIES MATCHES ".*\\.framework")
+      set(LIBR_HOME "${LIBR_LIBRARIES}/Resources")
+      set(LIBR_EXECUTABLE "${LIBR_HOME}/R")
+    else()
+      get_filename_component(_LIBR_LIBRARIES "${LIBR_LIBRARIES}" REALPATH)
+      get_filename_component(_LIBR_LIBRARIES_DIR "${_LIBR_LIBRARIES}" DIRECTORY)
+      set(LIBR_EXECUTABLE "${_LIBR_LIBRARIES_DIR}/../bin/R")
+    endif()
+
+  elseif(UNIX)
+
+    # attempt to find R executable
+    if(NOT LIBR_EXECUTABLE)
+      find_program(
+        LIBR_EXECUTABLE
+        NO_DEFAULT_PATH
+        HINTS "${CMAKE_CURRENT_BINARY_DIR}" "/usr/bin" "/usr/lib/" "/usr/local/bin/"
```

Comment:
Can you please explain why `"${CMAKE_CURRENT_BINARY_DIR}"` is here?

Reply:
I guess this could matter if you have fully self-contained apps where you are installing R and your application's code into the same location (or even if you are bundling the R executable alongside your code). That seems pretty non-standard but I guess it is possible?

Honestly I'm just speculating. I copied the use of this straight from `xgboost`'s setup. I left it because I thought "well it's working for them, maybe I shouldn't touch it". I am definitely ok dropping it if you'd like.

Reply:
OK, let's leave it as is. I was just curios about use-cases and  I thought you added this path intentionally.

Reply:
I think my "self-contained apps" thing is uncommon but it's possible. I have worked with systems before where on the only directory you have write access to is the one that your code runs in and its subdirectories.